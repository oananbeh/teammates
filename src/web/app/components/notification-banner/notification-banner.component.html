<div id="notification-banner" class="banner" [ngClass]="notifications[0].style | notificationStyleClass" *ngIf="isShown && notifications.length > 0" [attr.data-testid]="notifications[0].notificationId" @collapseAnim>
  <button id="btn-close-notif" type="button" class="btn-close" aria-label="Close" (click)="closeNotification()"></button>
  <h5>{{ notifications[0].title }}</h5>
  <div class="banner-text" [innerHTML]="notifications[0].message"></div>
  <div class="d-flex flex-row-reverse mt-3 me-2">
    <button id="btn-mark-as-read" type="button" [ngClass]="getButtonClass(notifications[0])" (click)="markNotificationAsRead(notifications[0])">
      Mark as Read
    </button>
  </div>
</div>
