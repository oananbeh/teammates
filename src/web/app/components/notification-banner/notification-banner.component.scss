.banner {
  padding: 1.875rem 1.875rem 1.25rem;
  word-break: break-word;
}

.banner-text {
  max-height: 80px;
  overflow-y: auto;
}

.btn-close {
  float: right;
}
