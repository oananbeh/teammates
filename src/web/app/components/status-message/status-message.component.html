<div class="alert alert-{{ statusMessage.color }}" *ngFor="let statusMessage of messages">
  <button class="btn-close" (click)="dismiss(statusMessage)"></button>
  <div style="padding-right: 30px;" [innerHtml]="statusMessage.message"></div>
</div>
