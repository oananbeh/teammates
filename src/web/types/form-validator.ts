import { ApiConst, ApiStringConst } from './api-const';

/**
 * Represents the root FormValidator object of all form fields.
 */
export enum FormValidator {
  /**
   * Max length for the 'Course ID' field.
   */
  COURSE_ID_MAX_LENGTH = ApiConst.COURSE_ID_MAX_LENGTH,

  /**
   * Max length for the 'Course Name' field.
   */
  COURSE_NAME_MAX_LENGTH = ApiConst.COURSE_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Student Name` field.
   */
  STUDENT_NAME_MAX_LENGTH = ApiConst.STUDENT_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Section Name` field.
   */
  SECTION_NAME_MAX_LENGTH = ApiConst.SECTION_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Team Name` field.
   */
  TEAM_NAME_MAX_LENGTH = ApiConst.TEAM_NAME_MAX_LENGTH,

  /**
   * Max length for the 'E-mail Address` field.
   */
  EMAIL_MAX_LENGTH = ApiConst.EMAIL_MAX_LENGTH,

  /**
   * Regex used to verify emails in the back-end.
   */
  EMAIL_REGEX = ApiStringConst.EMAIL_REGEX,

  /**
   * Regex used to verify names.
   *
   * Based on back-end's `FieldValidator.REGEX_NAME`.
   * The back-end regex is not converted to use here as the pattern syntax is not accepted in JS.
   */
  NAME_REGEX = '^[a-zA-Z0-9][^|%]*$',

  /**
   * Regex used to verify country names.
   *
   * Based on back-end's `FieldValidator.REGEX_NAME`, but without needing to start with alphanumeric
   * as the country is added to the end of the combined institute string.
   */
  COUNTRY_REGEX = '^[^|%]*$',

  /**
   * Max length for institution name in account request. (to be combined with country)
   */
  INSTITUTION_NAME_MAX_LENGTH = 86,

  /**
   * Max length for country in account request. (to be combined with institution name)
   */
  COUNTRY_NAME_MAX_LENGTH = 40,
}
