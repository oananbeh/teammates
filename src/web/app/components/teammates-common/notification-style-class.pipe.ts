import { Pipe, PipeTransform } from '@angular/core';
import { NotificationStyle } from '../../../types/api-output';

/**
 * Pipe to handle the transformation of an NotificationStyle to a css class.
 */
@Pipe({
  name: 'notificationStyleClass',
})
export class NotificationStyleClassPipe implements PipeTransform {

  transform(style: NotificationStyle): string {
    return `alert alert-${style.toLowerCase()}`;
  }

}
