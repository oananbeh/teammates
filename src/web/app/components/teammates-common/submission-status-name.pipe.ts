import { Pipe, PipeTransform } from '@angular/core';
import { DeadlineExtensionHelper } from '../../../services/deadline-extension-helper';
import { FeedbackSessionSubmissionStatus } from '../../../types/api-output';

/**
 * Pipe to handle the display of {@code FeedbackSessionSubmissionStatus}.
 */
@Pipe({
  name: 'submissionStatusName',
})
export class SubmissionStatusNamePipe implements PipeTransform {

  /**
   * Transforms {@link FeedbackSessionSubmissionStatus} to a simple name.
   */
  transform(status: FeedbackSessionSubmissionStatus, deadlines?: {
    studentDeadlines: Record<string, number>, instructorDeadlines: Record<string, number>,
  }): string {
    let string = '';
    switch (status) {
      case FeedbackSessionSubmissionStatus.NOT_VISIBLE:
      case FeedbackSessionSubmissionStatus.VISIBLE_NOT_OPEN:
        string += 'Awaiting';
        break;
      case FeedbackSessionSubmissionStatus.OPEN:
        string += 'Open';
        break;
      case FeedbackSessionSubmissionStatus.GRACE_PERIOD:
        string += 'Open (grace period)';
        break;
      case FeedbackSessionSubmissionStatus.CLOSED:
        string += 'Closed';
        break;
      default:
        return 'Unknown';
    }

    if (deadlines && DeadlineExtensionHelper.hasOngoingExtension(deadlines)) {
      string += ' (Ext. ongoing)';
    }

    return string;
  }
}
