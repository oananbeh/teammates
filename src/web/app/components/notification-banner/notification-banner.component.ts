import { Component, Input, OnChanges, OnInit } from '@angular/core';
import { NotificationService } from '../../../services/notification.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { Notification, Notifications, NotificationTargetUser } from '../../../types/api-output';
import { ErrorMessageOutput } from '../../error-message-output';
import { collapseAnim } from '../teammates-common/collapse-anim';

/**
 * Banner used to display notifications to the user.
 */
@Component({
  selector: 'tm-notification-banner',
  templateUrl: './notification-banner.component.html',
  styleUrls: ['./notification-banner.component.scss'],
  animations: [collapseAnim],
})
export class NotificationBannerComponent implements OnInit, OnChanges {

  @Input()
  url: string = '';

  @Input()
  notificationTargetUser: NotificationTargetUser = NotificationTargetUser.GENERAL;

  isShown: boolean = true;
  notifications: Notification[] = [];

  constructor(private notificationService: NotificationService,
              private statusMessageService: StatusMessageService) { }

  ngOnInit(): void {
    if (this.notificationTargetUser !== NotificationTargetUser.GENERAL) {
      this.fetchNotifications();
    }
  }

  ngOnChanges(): void {
    // Hide the notification banner if the user is on user notifications page
    if (this.url.includes('notifications')) {
      this.closeNotification();
    }
  }

  fetchNotifications(): void {
    this.notificationService.getUnreadNotificationsForTargetUser(this.notificationTargetUser)
      .subscribe((response: Notifications) => {
        this.notifications = response.notifications;
      });
  }

  markNotificationAsRead(notification: Notification): void {
    this.notificationService.markNotificationAsRead({
      notificationId: notification.notificationId,
      endTimestamp: notification.endTimestamp,
    })
      .subscribe({
        next: () => {
          this.statusMessageService.showSuccessToast('Notification marked as read.');
          this.closeNotification();
        },
        error: (resp: ErrorMessageOutput) => {
          this.statusMessageService.showErrorToast(resp.error.message);
        },
      });
  }

  closeNotification(): void {
    this.isShown = false;
  }

  getButtonClass(notification: Notification): string {
    return `btn btn-${notification.style.toLowerCase()}`;
  }
}
