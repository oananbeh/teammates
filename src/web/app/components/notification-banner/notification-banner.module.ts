import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NotificationBannerComponent } from './notification-banner.component';
import { TeammatesCommonModule } from '../teammates-common/teammates-common.module';

/**
 * Module for banner used to display notifications to the user.
 */
@NgModule({
  declarations: [NotificationBannerComponent],
  exports: [NotificationBannerComponent],
  imports: [CommonModule, TeammatesCommonModule],
})
export class NotificationBannerModule { }
