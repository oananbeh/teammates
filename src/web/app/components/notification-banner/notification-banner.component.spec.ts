import { CommonModule } from '@angular/common';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { of } from 'rxjs';
import SpyInstance = jest.SpyInstance;
import { NotificationBannerComponent } from './notification-banner.component';
import { NotificationService } from '../../../services/notification.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { Notification, NotificationStyle, NotificationTargetUser } from '../../../types/api-output';
import { MarkNotificationAsReadRequest } from '../../../types/api-request';
import { TeammatesCommonModule } from '../teammates-common/teammates-common.module';

describe('NotificationBannerComponent', () => {
  let component: NotificationBannerComponent;
  let statusMessageService: StatusMessageService;
  let notificationService: NotificationService;
  let fixture: ComponentFixture<NotificationBannerComponent>;

  const testNotificationOne: Notification = {
    notificationId: 'notification1',
    startTimestamp: new Date('2017-09-15T09:30+00:00').getTime(),
    endTimestamp: new Date('2050-09-15T09:30+00:00').getTime(),
    createdAt: new Date('2017-09-15T09:30+00:00').getTime(),
    style: NotificationStyle.SUCCESS,
    targetUser: NotificationTargetUser.GENERAL,
    title: 'valid title 1',
    message: 'valid message 1',
    shown: false,
  };

  const testNotificationTwo: Notification = {
    notificationId: 'notification2',
    startTimestamp: new Date('2018-12-15T09:30+00:00').getTime(),
    endTimestamp: new Date('2050-11-15T09:30+00:00').getTime(),
    createdAt: new Date('2018-11-15T09:30+00:00').getTime(),
    style: NotificationStyle.DANGER,
    targetUser: NotificationTargetUser.GENERAL,
    title: 'valid title 2',
    message: 'valid message 2',
    shown: false,
  };

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [NotificationBannerComponent],
      imports: [
        BrowserAnimationsModule,
        CommonModule,
        TeammatesCommonModule,
        HttpClientTestingModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(NotificationBannerComponent);
    notificationService = TestBed.inject(NotificationService);
    statusMessageService = TestBed.inject(StatusMessageService);
    component = fixture.componentInstance;
    fixture.detectChanges();
    component.notificationTargetUser = NotificationTargetUser.STUDENT;
    component.isShown = true;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should load data correctly', () => {
    const spy = jest.spyOn(notificationService, 'getUnreadNotificationsForTargetUser').mockReturnValue(of({
      notifications: [testNotificationOne, testNotificationTwo],
    }));
    component.ngOnInit();
    expect(spy).toHaveBeenCalledTimes(1);
    expect(component.notifications).toEqual([testNotificationOne, testNotificationTwo]);
  });

  it('should close after clicking X', () => {
    component.notifications = [testNotificationOne];
    fixture.detectChanges();

    expect(component.isShown).toBeTruthy();
    const button = fixture.debugElement.query(By.css('#btn-close-notif')).nativeElement;
    button.click();
    expect(component.isShown).toBeFalsy();
  });

  it('should close after clicking mark as read', () => {
    const apiSpy: SpyInstance = jest.spyOn(notificationService, 'markNotificationAsRead')
      .mockImplementation((request: MarkNotificationAsReadRequest) => {
        expect(request.notificationId).toEqual(testNotificationOne.notificationId);
        return of({
          readNotifications: [request.notificationId],
        });
      });
    const messageSpy: SpyInstance = jest.spyOn(statusMessageService, 'showSuccessToast')
      .mockImplementation((args: string) => {
        expect(args).toEqual('Notification marked as read.');
      });
    component.notifications = [testNotificationOne];
    fixture.detectChanges();

    expect(component.isShown).toBeTruthy();
    const button = fixture.debugElement.query(By.css('#btn-mark-as-read')).nativeElement;
    button.click();
    expect(apiSpy).toHaveBeenCalledTimes(1);
    expect(messageSpy).toHaveBeenCalledTimes(1);
    expect(component.isShown).toBeFalsy();
  });

  it('should snap with no unread notifications', () => {
    component.notifications = [];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with 1 unread notification', () => {
    component.notifications = [testNotificationOne];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with multiple unread notifications', () => {
    component.notifications = [testNotificationOne, testNotificationTwo];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
