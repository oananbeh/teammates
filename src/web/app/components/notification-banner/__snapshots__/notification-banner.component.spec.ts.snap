// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationBannerComponent should snap with 1 unread notification 1`] = `
<tm-notification-banner
  isShown={[Function Boolean]}
  notificationService={[Function NotificationService]}
  notificationTargetUser={[Function String]}
  notifications={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  url=""
>
  <div
    class="banner ng-trigger ng-trigger-collapseAnim alert alert-success ng-animating"
    data-testid="notification1"
    id="notification-banner"
  >
    <button
      aria-label="Close"
      class="btn-close"
      id="btn-close-notif"
      type="button"
    />
    <h5>
      valid title 1
    </h5>
    <div
      class="banner-text"
    >
      valid message 1
    </div>
    <div
      class="d-flex flex-row-reverse mt-3 me-2"
    >
      <button
        class="btn btn-success"
        id="btn-mark-as-read"
        type="button"
      >
         Mark as Read 
      </button>
    </div>
  </div>
</tm-notification-banner>
`;

exports[`NotificationBannerComponent should snap with multiple unread notifications 1`] = `
<tm-notification-banner
  isShown={[Function Boolean]}
  notificationService={[Function NotificationService]}
  notificationTargetUser={[Function String]}
  notifications={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  url=""
>
  <div
    class="banner ng-trigger ng-trigger-collapseAnim alert alert-success ng-animating"
    data-testid="notification1"
    id="notification-banner"
  >
    <button
      aria-label="Close"
      class="btn-close"
      id="btn-close-notif"
      type="button"
    />
    <h5>
      valid title 1
    </h5>
    <div
      class="banner-text"
    >
      valid message 1
    </div>
    <div
      class="d-flex flex-row-reverse mt-3 me-2"
    >
      <button
        class="btn btn-success"
        id="btn-mark-as-read"
        type="button"
      >
         Mark as Read 
      </button>
    </div>
  </div>
</tm-notification-banner>
`;

exports[`NotificationBannerComponent should snap with no unread notifications 1`] = `
<tm-notification-banner
  isShown={[Function Boolean]}
  notificationService={[Function NotificationService]}
  notificationTargetUser={[Function String]}
  notifications={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  url=""
>
</tm-notification-banner>
`;
